File: Hash/242.有效的字母异位词.cpp
#include <iostream>
#include <string>

using namespace std;
/*
 * @lc app=leetcode.cn id=242 lang=cpp
 *
 * [242] 有效的字母异位词
 */

// @lc code=start
class Solution
{
public:
    bool isAnagram(string s, string t)
    {
        if (s.size() != t.size())
        {
            return false;
        }

        // 记录每个字母出现次数的数组
        int arr[26] = {0};
        for (int i = 0; i < s.size(); i++)
        {
            // 'a' - 'a' = 0，'b' - 'a' = 1
            // 实际上是ascii码相减，所以可以作为数组下标
            arr[s[i] - 'a']++;
            arr[t[i] - 'a']--;
        }

        for (int i = 0; i < 26; i++)
        {
            if (arr[i] != 0)
            {
                return false;
            }
        }

        return true;
    }
};
// @lc code=end

File: Hash/242.有效的字母异位词.ts
/*
 * @lc app=leetcode.cn id=242 lang=typescript
 *
 * [242] 有效的字母异位词
 */

// @lc code=start
function isAnagram(s: string, t: string): boolean {
  let result = {}
  if (!s || !t) {
    return false
  }

  if (s.length !== t.length) {
    return false
  }

  // 使用对象记录出现次数
  for (let i = 0; i < s.length; i++) {
    result[s[i]] = (result[s[i]] || 0) + 1
    result[t[i]] = (result[t[i]] || 0) - 1
  }

  // 对象中有不为0的，说明这个字母在两个字符串中出现次数不一样
  for (let key in result) {
    if (result[key] !== 0) {
      return false
    }
  }

  return true
};
// @lc code=end

